### PR TITLE
Refactor platform message logic

### PR DESCRIPTION
**Pull request by Hixie**
This laids the groundwork for sending messages through ChannelBuffers with channel-specific callbacks rather than a single onPlatformMessage callback.

This allows us to remove the logic from the framework that puts data back into the channel buffers. Right now (before this PR) the logic for messages from plugins to the framework is bidirectional:

```
                  **                 *
Plugins -> Engine -> ChannelBuffers <- Framework <---+-.
               |              |                      | |
               |              '------> via drain ----' |
               |                                       |
               '----------------- onPlatformMessage ---'

* = when the listener is null on the framework side
** = when onPlatformMessage is null
```

This ends up with weird race conditions and is generally less than completely clear. With this PR, we lay the groundwork for eventually reaching this model:

```
Plugins -> Engine -> ChannelBuffers -> Framework
```

...which is significantly simpler.

### CODE REVIEW

**Comment by Hixie** [comment]
I'm expecting tests to fail, I didn't run them locally.

**Comment by Hixie** [comment]
Fixes https://github.com/flutter/flutter/issues/66699.

File: lib/ui/channel_buffers.dart [hunk 1]
```diff
@@ -6,184 +6,334 @@
...
-      queue = _makeRingBuffer(kDefaultBufferSize);
-      _messages[channel] = queue;
-    }
-    final bool didOverflow = queue.push(_StoredMessage(data, callback));
-    if (didOverflow) {
-      // TODO(gaaclarke): Update this message to include instructions on how to resize
-      // the buffer once that is available to users and print in all engine builds
-      // after we verify that dropping messages isn't part of normal execution.
-      _printDebug('Overflow on channel: $channel. '
-                  'Messages on this channel are being discarded in FIFO fashion. '
```
Comment (gaaclarke) [inline lib/ui/channel_buffers.dart:144]: You can't get rid of this warning.  We can't just drop messages without warning developers.  It's particularly an issue with MethodChannels that multiplex multiple messages onto one channel.

File: lib/ui/hooks.dart [hunk 2]
```diff
@@ -266,6 +267,22 @@ void _invoke1<A>(void callback(A a)?, Zone zone, A arg) {
   }
 }
 
+/// Invokes [callback] inside the given [zone] passing it [arg1] and [arg2].
```
Comment (gaaclarke) [inline lib/ui/hooks.dart:270]: nit: I'd note arity 2 inside the docstring to immediately communicate the meaning of 2 in the name.

File: lib/ui/channel_buffers.dart [hunk 3]
```diff
@@ -6,184 +6,334 @@
...
+    _channelCallbackRecord = _ChannelCallbackRecord(callback);
+    if (needDrain && !_draining)
+      _drain();
+  }
+
+  /// Clears the listener for this channel.
+  ///
+  /// When there is no listener, messages are queued, up to [capacity],
+  /// and then discarded in a first-in-first-out fashion.
+  void clearListener() {
```
Comment (gaaclarke) [inline lib/ui/channel_buffers.dart:172]: Clearing the listener while draining will lead to a null pointer exception.  I thought we said we would drop the messages.

File: lib/ui/channel_buffers.dart [hunk 4]
```diff
@@ -6,184 +6,334 @@
...
   /// This should be called once a channel is prepared to handle messages
   /// (i.e. when a message handler is setup in the framework).
-  Future<void> drain(String channel, DrainChannelCallback callback) async {
-    while (!_isEmpty(channel)) {
-      final _StoredMessage message = _pop(channel)!;
+  ///
+  /// The messages are processed by calling the given `callback`. Each message
+  /// is processed in its own microtask.
+  // TODO(ianh): deprecate once framework uses [setListener].
+  Future<void> drain(String name, DrainChannelCallback callback) async {
```
Comment (gaaclarke) [inline lib/ui/channel_buffers.dart:333]: I'm not sure that anyone ever took advantage of the future that was returned here, but now there is no way for a client to know when the draining process is done with the setListener API.  There isn't a place to put it either.  I would rather keep the explicit drain call, maybe replace this with `_drainAsynchronously`?

File: lib/ui/channel_buffers.dart [hunk 5]
```diff
@@ -6,184 +6,334 @@
...
+    assert(_draining);
+    if (_queue.isNotEmpty && _channelCallbackRecord != null) {
+      final _StoredMessage message = pop();
+      _channelCallbackRecord!.invoke(message.data, message.callback);
+      scheduleMicrotask(_drainStep);
+    } else {
+      _draining = false;
+    }
+  }
+}
```
Comment (gaaclarke) [inline lib/ui/channel_buffers.dart:209]: The previous implementation of drain would handle the messages serially, this will potentially parallelize asynchronous operations.  I think that is good since that behavior will match channels when they aren't buffering, correct?

**Review by gaaclarke — COMMENTED** [review]
This looks good for the most part the biggest points I have are:
1) The behavior of clearing the listener while draining
2) Never drop messages without telling the developer, bugs created by this are difficult to debug
3) An explicit call to drain gives us a place to put logic that happens after the drain

File: lib/ui/channel_buffers.dart (hunk 4 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:333]: I'd just say launching any asynchronous operation without returning a handle to that operation is a red flag since it hides the fact that something is asynchronous and provides no mechanism for someone to synchronize if they want to.

File: lib/ui/channel_buffers.dart (hunk 1 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:144]: My understanding is that dropping messages here is part of the expected behaviour of the API. Is that not correct? The documentation talks about how you can legitimately want to only be holding on to the newest message (e.g. because you're a level-trigger API).

We have a policy of only logging messages that are actionable (see style guide). If the messages are not actionable, we shouldn't log the message. I don't know what action one should take with this message.

I agree that if you're dropping messages when you don't expect to be (e.g. you're listening to the wrong channel) then this message would make debugging the issue much easier. Is there some way we can determine when the message is actionable and when it's just logspam?

File: lib/ui/hooks.dart (hunk 2 above)
Comment (Hixie) [reply lib/ui/hooks.dart:270]: added

File: lib/ui/channel_buffers.dart (hunk 3 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:172]: I will add a test for this case.

My intention was for removing a listener to just cause the messages to stop being drained, without losing them.

File: lib/ui/channel_buffers.dart (hunk 4 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:333]: Right now the drain API only has one client, the framework, and it ignores the future.

That said, I think that exposing the end of the drain is a dangerous affordance. The point of this API is to buffer the messages so that when the plugin is ready to receive them, it gets them. The sender has no way to know if the plugin is already listening or not when it sends the message, so by definition the messages it sends cannot depend on whether or not the plugin is listeneing. If we allow messages received before the plugin subscribes to be distinguished to those received after the plugin subscribes, we allow the plugin to behave differently for events sent before and after subscription, but since we just said the sender is not distinguishing, that implies that we are allowing behavior variance along an axis that should have none (i.e., it would be a bug vector).

It also would encourage race conditions. Consider these two cases:

A:
* t=0ms plugin requests buffer size 3
* t=0ms message 1
* t=2ms message 2
* t=3ms plugin subscribes
* t=4ms message 3

B:
* t=0ms plugin requests buffer size 3
* t=0ms message 1
* t=1ms plugin subscribes
* t=2ms message 2
* t=4ms message 3

Why would we want the plugin to react differently to message 2 in the first case vs the second case?

To put it another way, what's the use case for knowing when the drain is complete?

File: lib/ui/channel_buffers.dart (hunk 5 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:209]: The callbacks are synchronous, so they are guaranteed to be complete before the next microtask is scheduled. This matches the way messages are handled outside of drain.

File: lib/ui/channel_buffers.dart (hunk 1 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:144]: Nope, it's not expected.  Running the engine is an asynchronous operation and we don't provide any mechanism for users to determine if the engine is running before we can send messages.

This code will fail to deliver a message randomly:
```
[engine run];
usleep(rand() * 1000000);
[channel sendMessage:@"hello"];
```

One action that could fix the problem is to chose a more appropriate buffer size, which is mentioned.  If you are using method channels you'd probably want to base that calculus off of the number of unique messages you are sending over that channel.

It also mentions that the engine isn't running which admittedly is a hint about the asynchronous nature of running engines.  With that information they could solve the problem a few other ways.

When I first implemented channel buffers we saw there were all kinds of messages routinely being dropped during the initialization of an engine.  They were not expected.

File: lib/ui/channel_buffers.dart (hunk 3 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:172]: test seems to confirm that clearListener() just pauses the draining.

Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:172]: Ahh yep, thanks I missed the null check before its usage.

File: lib/ui/channel_buffers.dart (hunk 5 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:209]: Yea, they might spawn asynchronous operations but I think it was erroneous to wait for those operations previously.  This sounds good to me.

File: lib/ui/channel_buffers.dart (hunk 4 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:333]: I admit it is hypothetical.  Hiding an asynchronous operation in a synchronous signature can cause bugs.  It's not as helpful in Dart though because Dart allows you to drop return values implicitly.

Someday someone might write this code:
```dart
_drain();
clearListener();
```

This is a nonsensical operation because the clearListener would immediately disable the drain before anything happens.

If `_drain` advertises itself as an asynchronous operation by returning a `Future<void>` that mistake is less likely.  I don't know what code some future developer might write but I'd error on the side of giving them best hope of writing correct code.

File: lib/ui/channel_buffers.dart [hunk 6]
```diff
@@ -6,184 +6,334 @@
...
+  /// previous one.
+  ///
+  /// Callbacks are invoked in their own stack frame and
+  /// use the zone that was current when the callback was
+  /// registered.
+  ///
+  /// ## Draining
+  ///
+  /// If any messages were queued before the listener is added,
+  /// they are drained asynchronously after this method returns.
```
Comment (yjbanov) [inline lib/ui/channel_buffers.dart:300]: Consider linking to the `drain` method at the end of the paragraph. A reader may not know what "drain" means.

File: lib/ui/channel_buffers.dart [hunk 7]
```diff
@@ -6,184 +6,334 @@
...
   ///
   /// This should be called once a channel is prepared to handle messages
   /// (i.e. when a message handler is setup in the framework).
-  Future<void> drain(String channel, DrainChannelCallback callback) async {
-    while (!_isEmpty(channel)) {
-      final _StoredMessage message = _pop(channel)!;
+  ///
+  /// The messages are processed by calling the given `callback`. Each message
+  /// is processed in its own microtask.
+  // TODO(ianh): deprecate once framework uses [setListener].
```
Comment (yjbanov) [inline lib/ui/channel_buffers.dart:332]: If this method is going away we should explain the term "drain" in other dartdocs.

File: lib/ui/hooks.dart [hunk 8]
```diff
@@ -253,6 +254,8 @@ void _invoke(void callback()?, Zone zone) {
 }
 
 /// Invokes [callback] inside the given [zone] passing it [arg].
+///
+/// The 1 in the name refers to the number of arguments expected by the callback.
```
Comment (yjbanov) [inline lib/ui/hooks.dart:258]: ...as well as the number of arguments passed to this function, not counting `callback` and `zone`.

File: lib/web_ui/lib/src/engine/window.dart [hunk 9]
```diff
@@ -478,14 +480,30 @@ class EngineWindow extends ui.Window {
...
+    String name,
+    ByteData? data,
+    ui.PlatformMessageResponseCallback callback,
+  ) {
+    if (name == ui.ChannelBuffers.kControlChannelName) {
+      // TODO(ianh): move this logic into ChannelBuffers once we remove onPlatformMessage
+      try {
+        channelBuffers.handleMessage(data!);
+      } catch (ex) {
+        print('Message to "$name" caused exception $ex');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart:492]: This should probably log to https://api.dart.dev/stable/2.10.1/dart-html/Console/error.html, and should also include the stack trace.

**Review by yjbanov — COMMENTED** [review]
/cc @mdebbar @nturgut for changes in navigation and text input.

File: testing/dart/channel_buffers_test.dart [hunk 10]
```diff
@@ -3,6 +3,8 @@
 // found in the LICENSE file.
 
 // @dart = 2.6
+
```
Comment (yjbanov) [inline testing/dart/channel_buffers_test.dart:6]: We currently do not run these tests against the web version and instead maintain separate tests. Can you please copy this test into lib/web_ui/test?

/cc @nturgut how feasible is it to run these tests against the web version? One difference between these tests and ours is that these tests input `dart:ui`, while web tests import `package:ui/ui.dart`.

File: lib/ui/channel_buffers.dart (hunk 1 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:144]: Ok, we should fix the documentation then, because right now it clearly states that there are cases where dropping messages is the appropriate behavior.

Maybe we should show the message for _any_ message that is buffered, if the buffer size hasn't been set? I guess in that case we probably would want to default to a buffer size of zero, too, since we'd be saying all plugins need to set their buffer size if they send messages before the framework is ready.

File: lib/ui/channel_buffers.dart (hunk 4 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:333]: Sorry for not being clear. My medium term plan is to remove the drain API, it just becomes an implicit behavior of setListener. The idea is to remove the concept of draining from the API -- it just behaves exactly as if messages were only sent while the listener was set, and they get processed the same as if there were messages in fact only being sent while the listener was set.

File: lib/ui/channel_buffers.dart [hunk 11]
```diff
@@ -6,184 +6,334 @@
...
+///
+/// This consists of a fixed-size circular queue of [_StoredMessage]s,
+/// and the channel's callback, if any has been registered.
+class _Channel {
+  _Channel(this._capacity)
+    : _queue = collection.ListQueue<_StoredMessage>(_capacity);
 
-  _RingBuffer(this._capacity)
-    : _queue = collection.ListQueue<T>(_capacity);
+  /// The underlying data for the RingBuffer.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:70]: fixed

File: lib/ui/channel_buffers.dart [hunk 12]
```diff
@@ -6,184 +6,334 @@
...
-  Function(T)? _dropItemCallback;
-  set dropItemCallback(Function(T) callback) {
-    _dropItemCallback = callback;
+  /// Set the [capacity] of the channel to the given size.
+  ///
+  /// If the new size is smaller than the [length], the oldest
+  /// messages are discarded until the capacity is reached.
+  ///
+  /// Returns the number of discarded messages resulting from resize.
+  int resize(int newSize) {
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:93]: done

File: lib/ui/channel_buffers.dart [hunk 13]
```diff
@@ -6,184 +6,334 @@
...
+/// and the channel's callback, if any has been registered.
+class _Channel {
+  _Channel(this._capacity)
+    : _queue = collection.ListQueue<_StoredMessage>(_capacity);
 
-  _RingBuffer(this._capacity)
-    : _queue = collection.ListQueue<T>(_capacity);
+  /// The underlying data for the RingBuffer.
+  ///
+  /// [ListQueue]s dynamically resize, but [_Channel]s do not.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:72]: removed this sentence

File: lib/ui/channel_buffers.dart [hunk 14]
```diff
@@ -6,184 +6,334 @@
...
-  /// by way of an overflow or a resizing.
-  Function(T)? _dropItemCallback;
-  set dropItemCallback(Function(T) callback) {
-    _dropItemCallback = callback;
+  /// Set the [capacity] of the channel to the given size.
+  ///
+  /// If the new size is smaller than the [length], the oldest
+  /// messages are discarded until the capacity is reached.
+  ///
+  /// Returns the number of discarded messages resulting from resize.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:92]: Since this is now a setter, this line has been removed.

File: lib/ui/channel_buffers.dart [hunk 15]
```diff
@@ -6,184 +6,334 @@
...
+    _draining = true;
+    scheduleMicrotask(_drainStep);
   }
-}
 
-/// Signature for [ChannelBuffers.drain].
-typedef DrainChannelCallback = Future<void> Function(ByteData?, PlatformMessageResponseCallback);
+  /// Drains a single message and then reinvokes itself asynchronously.
+  ///
+  /// See [drain] for more details.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:198]: done

File: lib/ui/channel_buffers.dart [hunk 16]
```diff
@@ -6,184 +6,334 @@
...
+    } else {
+      _draining = false;
+    }
+  }
+}
 
-/// Storage of channel messages until the channels are completely routed,
-/// i.e. when a message handler is attached to the channel on the framework side.
+/// The buffering and dispatch mechanism for messages sent by plugins to
+/// the Dart code.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:212]: fixed

File: lib/ui/channel_buffers.dart [hunk 17]
```diff
@@ -6,184 +6,334 @@
...
 class ChannelBuffers {
-  /// By default we store one message per channel.  There are tradeoffs associated
-  /// with any size.  The correct size should be chosen for the semantics of your
-  /// channel.
+  /// Create a buffer pool for platform messages.
   ///
-  /// Size 0 implies you want to ignore any message that gets sent before the engine
-  /// is ready (keeping in mind there is no way to know when the engine is ready).
+  /// It is generally not necessary to create an instance of this class,
+  /// the global [channelBuffers] instance is the one used by the engine.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:226]: replaced comma with semicolon.

File: lib/ui/channel_buffers.dart [hunk 18]
```diff
@@ -198,25 +348,46 @@ class ChannelBuffers {
...
+  /// **Format:** `resize\r<channel name>\r<new size>`
+  ///
+  /// **Description:** Allows you to set the size of a channel's buffer.
   void handleMessage(ByteData data) {
     final List<String> command = _getString(data).split('\r');
-    if (command.length == /*arity=*/2 + 1 && command[0] == 'resize') {
+    if (command.length == 1 + /*arity=*/2 && command[0] == 'resize') {
       _resize(command[1], int.parse(command[2]));
     } else {
       throw Exception('Unrecognized command $command sent to $kControlChannelName.');
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:367]: replaced command with message.

File: lib/ui/channel_buffers.dart [hunk 19]
```diff
@@ -6,184 +6,334 @@
...
-    return result;
-  }
-
-  void _onDropItem(_StoredMessage message) {
-    message.callback(null);
-  }
+  /// A mapping between a channel name and its associated [_Channel].
+  final Map<String, _Channel> _channels = <String, _Channel>{};
 
+  /// Adds a message (`data`) to the named channel buffer (`channel`).
```
[the remark this answers is not in the document]
Comment (Hixie) [reply lib/ui/channel_buffers.dart:274]: fixed

File: lib/ui/channel_buffers.dart (hunk 6 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:300]: done

File: lib/ui/channel_buffers.dart (hunk 7 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:332]: The docs for setListener have a section on this.

File: lib/ui/hooks.dart (hunk 8 above)
Comment (Hixie) [reply lib/ui/hooks.dart:258]: added

File: lib/web_ui/lib/src/engine/window.dart (hunk 9 above)
Comment (Hixie) [reply lib/web_ui/lib/src/engine/window.dart:492]: Yeah I was wondering about this. Maybe we shouldn't even catch the exception here, and let the usual uncaught exception logic do its thing? I've removed the catch entirely for now. cc @gaaclarke who added this catch originally.

File: testing/dart/channel_buffers_test.dart (hunk 10 above)
Comment (Hixie) [reply testing/dart/channel_buffers_test.dart:6]: I had actually already ported the tests but forgot to `git add` them, oops.

File: lib/ui/channel_buffers.dart (hunk 1 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:144]: Another option would be to leave the default as now, but have a specific control message to silence the message (and have the warning log say to send this control message).

Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:144]: Can you point out the documentation that says dropping messages is appropriate?  One other caveat we haven't discussed is that there are 2 types of channels and the semantics aren't explicit, channels where dropping messages is fine (think sending the gyro readings to the engine) and channels where it isn't (think sending entries in a database).  In other words, is the message interpreted as `set` or `add`?  Is the documentation referring to that?

Buffer size zero by default would be a big breaking change.  I wouldn't recommend that.

The worse case scenario is that users are losing data when they don't expect it.  The print statement is meant to warn users of that scenario.  I worry moving the print statement out of that scenario may cause it to be spammy and less likely heeded when it is appropriate.

However we slice things, not printing out the error despite having set the buffer size wouldn't be ideal since you'd have no idea that you may have chosen poorly for your use case.

> since we'd be saying all plugins need to set their buffer size if they send messages before the framework is ready.  

Remember there is no way to know when the framework is ready clients can't meaningfully know that they may be doing something that may not work.

> have a specific control message to silence the message

That's cool with me.  A bigger change would be if we could know the semantics of the channel and only print out the error if it has `add` semantics instead of `set` semantics.  The ship might have sailed on that though.

File: lib/ui/channel_buffers.dart (hunk 4 above)
Comment (gaaclarke) [reply lib/ui/channel_buffers.dart:330]: Yea, I got that.  I think it's a worse design because it's hiding asynchronous logic inside of the setter.  This is hiding side effects and temporal dependencies which is a liability.  If you are keen on this design naming `setListener` to `setListenerAndDrain` would be an improvement.

File: lib/ui/channel_buffers.dart (hunk 1 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:144]: https://master-api.flutter.dev/flutter/dart-ui/ChannelBuffers/kDefaultBufferSize-constant.html is the documentation I was talking about (I expand on it in this PR). It says "Size 1 implies that you only care about the most recent value." and says size 1 is the default, which implies that dropping messages is fine by default.

We could improve the control channel API in general, making it support a way to specify if it's being used for an event sequence (where dropping is bad) or where it's being used as a level-trigger (where dropping is fine). That would also give us the opportunity to make it use the standard binary format instead of its current inefficent bespoke text format with '\r' as the separator. Let me see what I can do.

**Review by mdebbar — APPROVED** [review]
Changes related to keyboard, history, text and text editing look good to me. The rest was covered by @yjbanov's review I believe.

**Comment by nturgut** [comment]
> Changes related to keyboard, history, text and text editing look good to me. The rest was covered by @yjbanov's review I believe.

our integration tests will fail if text_editing breaks :) They are testing almost every command that passes through platform channels.

**Comment by nturgut** [comment]
I'm a little confused, I have a question, if the end goal is this and we want to remove logic "data back into the channel buffers"

```
Plugins -> Engine -> ChannelBuffers -> Framework
```

what changes we need for [clipboard](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/clipboard.dart#L55)(The equivalent web engine code would [this]( https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/clipboard.dart) )

File: lib/web_ui/test/channel_buffers_test.dart [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' as ui;
+
+import 'package:test/bootstrap/browser.dart';
+import 'package:test/test.dart';
+
+void main() {
```
Comment (nturgut) [inline lib/web_ui/test/channel_buffers_test.dart:17]: our tests needs a different main function [ex](https://github.com/flutter/engine/blob/master/lib/web_ui/test/engine/surface/scene_builder_test.dart#L22)

**Comment by Hixie** [comment]
> I'm a little confused, I have a question, if the end goal is this and we want to remove logic "data back into the channel buffers"
> 
> ```
> Plugins -> Engine -> ChannelBuffers -> Framework
> ```
> 
> what changes we need for [clipboard](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/clipboard.dart#L55)(The equivalent web engine code would [this](https://github.com/flutter/engine/blob/master/lib/web_ui/lib/src/engine/clipboard.dart) )

The change will be entirely transparent to plugins. It only affects a little bit of code deep in the services library.

File: lib/ui/channel_buffers.dart (hunk 4 above)
Comment (Hixie) [reply lib/ui/channel_buffers.dart:381]: I don't really understand the concern here. Can you elaborate?

The point of this design is to make the draining entirely transparent, indistinguishable from not draining. There should be no way to distinguish someone connecting to a plugin before the plugin sends messages from connecting after the plugin sends messages. There's no "draining" anymore, conceptually, it's just that the messages weren't delivered until there was a listener.

**Comment by chinmaygarde** [comment]
The presubmit failures look like linter issues. Can you please fix and land this patch?

**Review by yjbanov — APPROVED** [review]
LGTM if LGT @gaaclarke 

lgtm

**Comment by gaaclarke** [comment]
Yea LGTM.  The big thing for me was keeping the warning on overflow, thanks.

**Comment by Hixie** [comment]
(i'll repush when the tree failure is fixed)

**Comment by chinmaygarde** [comment]
@Hixie Tree is open now. Let's land this ASAP before more conflicts creep in.

**Comment by chinmaygarde** [comment]
I see he is out this week. Will followup when he returns.

**Comment by Hixie** [comment]
I was going to try to land this but given the conflicts it'll have to take longer since I'll have to figure out how to merge it in.

**Comment by chinmaygarde** [comment]
Sounds good. I'll add the WIP tag so I don't keep nagging you during PR triage.

**Comment by Hixie** [comment]
I'll close this in favour of a merged version here: https://github.com/flutter/engine/pull/22181

**Comment by Hixie** [comment]
Ok, the merge is ready in https://github.com/flutter/engine/pull/22181. It can land once it has a review.
